apm-agent-plugins/apm-scala-concurrent-plugin/src/test/scala/co/elastic/apm/agent/concurrent/FutureInstrumentationSpec.scala

```diff
@@ -0,0 +1,64 @@
+package co.elastic.apm.agent.concurrent
+
+import java.util.concurrent.Executors
+
+import co.elastic.apm.agent.AbstractInstrumentationTest
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+
+import scala.concurrent.duration._
+import scala.concurrent.{Await, ExecutionContext, Future}
+
+class FutureInstrumentationSpec extends AbstractInstrumentationTest {
+
+  @Test
+  def testWithDefaultConfig(): Unit = {
+    new TestFutureTraceMethods().invokeAsync()
+    assertThat(AbstractInstrumentationTest.reporter.getTransactions().toArray).hasSize(1)
+    assertThat(AbstractInstrumentationTest.reporter.getSpans().toArray).hasSize(4)
+  }
+
+  private class TestFutureTraceMethods {
+
+    /**
+     * Calling this method results in this method call tree:
+     *
+     *                      main thread                         |           worker thread
+     * -------------------------------------------------------------------------------------------
+     * invokeAsync                                              |
+     *      |                                                   |
+     *      --- blockingMethodOnMainThread                      |
+     *                     |                                    |
+     *                     --- nonBlockingMethodOnMainThread    |
+     *                                      |                   |
+     *                                      --------------------------> methodOnWorkerThread
+     *                                                          |                |
+     *                                                          |                --- longMethod
+     *                                                          |
+     */
+    def invokeAsync(): Unit = blockingMethodOnMainThread()
+
+    private def blockingMethodOnMainThread(): Unit = {
+      try {
+        Await.result(nonBlockingMethodOnMainThread(), 10.seconds)
+      } catch {
+        case e: Exception => e.printStackTrace()
+      }
+    }
+
+    private def nonBlockingMethodOnMainThread(): Future[Unit] =
+      Future(methodOnWorkerThread())(ExecutionContext.fromExecutor(Executors.newFixedThreadPool(1)))
```

Comment:
Maybe here's something going wrong with the instrumentation. One limitation of the unit tests is that you can't test instrumenting classes from the bootstrap class loader. `ThreadPoolExecutor`, which is returned from `Executors.newFixedThreadPool(1)` is such a class.

To work around this, we use `co.elastic.apm.agent.concurrent.ExecutorServiceWrapper` in our tests. This is a class that the agent can instrument in tests that just delegates to the wrapped `ExecutorService`.

For the built-in executor service instrumentation to work, you also have to declare a test dependency to the java-concurrent-plugin (see comment on pom.xml).

To access the `ExecutorServiceWrapper` class, you either need to copy/paste it to your module or build a test jar from the java-concurrent-plugin and add it as a test dependency as well (see comment on pom.xml).